Fix `VecDeque::shrink_to` UB when `handle_alloc_error` unwinds.

Sp00ph opened the pull request:
Fixes #123369

For `VecDeque` it's relatively simple to restore the buffer into a consistent state so this PR does just that.

Note that with its current implementation, `shrink_to` may change the internal arrangement of elements in the buffer, so e.g. `[D, <uninit>, A, B, C]` will become `[<uninit>, A, B, C, D]` and `[<uninit>, <uninit>, A, B, C]` may become `[B, C, <uninit>, <uninit>, A]` if `shrink_to` unwinds. This shouldn't be an issue though as we don't make any guarantees about the stability of the internal buffer arrangement (and this case is impossible to hit on stable anyways).

This PR also includes a test with code adapted from #123369 which fails without the new `shrink_to` code. Does this suffice or do we maybe need more exhaustive tests like in #108475?

cc @Amanieu 

@rustbot label +T-libs

Sp00ph committed:
Fix `VecDeque::shrink_to` UB when `handle_alloc_error` unwinds.

Luckily it's comparatively simple to just restore the `VecDeque` into a valid state on unwinds.

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit fb00ac614406658fb263ba613b275416e90710c9 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum fb00ac614406658fb263ba613b275416e90710c9 -->
<!-- homu: {"type":"Approved","sha":"fb00ac614406658fb263ba613b275416e90710c9","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->

matthiaskrgr commented:
@bors r- https://github.com/rust-lang/rust/pull/123910#issuecomment-2053937742

MarkusEverling commented:
Hmm, so is this just blocked on rust-lang/miri#3439 then? Or should I just slap a `#[cfg(not(miri))]` on the test for now?

Sp00ph commented:
Woops, didn't mean to comment here from that account lol.

RalfJung commented on library/alloc/src/collections/vec_deque/tests.rs:
```diff
@@ -790,6 +797,52 @@ fn test_shrink_to() {
+
+    // preserve the old error hook just in case.
+    let old_error_hook = take_alloc_error_hook();
+    set_alloc_error_hook(|_| panic!("alloc error"));
```
This changes the alloc error hook for all concurrently running tests... is that really a good idea?

Sp00ph replied:
Hm, is there another mechanism to allow what I want to test there? Something like compiling this specific test with `-Zoom=unwind` maybe?

RalfJung replied:
You can make it a separate test crate by adding a file like `alloc/tests/vec_deque_alloc_error.rs` and add it to `alloc/Cargo.toml` separately.

RalfJung commented:
> Hmm, so is this just blocked on [rust-lang/miri#3439](https://github.com/rust-lang/miri/issues/3439) then? Or should I just slap a `#[cfg(not(miri))]` on the test for now?

Yeah disabling the test in Miri is fine for now, but please ping `@rust-lang/miri` when you do that so we can keep tabs of these tests and re-enable them when the underlying issue is fixed.

Sp00ph committed:
Move `test_shrink_to_unwind` to its own file.

This way, no other test can be tripped up by `test_shrink_to_unwind` changing the alloc error hook.

Sp00ph commented:
@rustbot ready

Mark-Simulacrum commented:
@bors r+

bors commented:
:pushpin: Commit 5cb53bc34d462a59629e5430cf2e29fe6820c550 has been approved by `Mark-Simulacrum`

It is now in the [queue](https://bors.rust-lang.org/homu/queue/rust) for this repository.

<!-- @bors r=Mark-Simulacrum 5cb53bc34d462a59629e5430cf2e29fe6820c550 -->
<!-- homu: {"type":"Approved","sha":"5cb53bc34d462a59629e5430cf2e29fe6820c550","approver":"Mark-Simulacrum","queue":"https://bors.rust-lang.org/homu/queue/rust"} -->
